### ISSUE TITLE
Conditional render is nested wrong using Rails 7.1

### ISSUE THREAD

**Issue by manufaktor** [haml/haml#1153]
Edit: Using haml-6.2.3
I've discovered the following problem after upgrading to Rails 7.1:

**Given the following HAML application layout in Rails 7.0**

```haml
!!!
%html.h-full.bg-gray-100
  %head
    %meta{:content => "text/html; charset=UTF-8", "http-equiv" => "Content-Type"}/
    %title Plannerapp
    %meta{:content => "width=device-width,initial-scale=1", :name => "viewport"}/
    = csrf_meta_tags
    = csp_meta_tag
    = stylesheet_link_tag "application", "data-turbo-track": "reload"
    = javascript_include_tag "application", "data-turbo-track": "reload", defer: true
  %body.h-full
    .min-h-full
      = turbo_frame_tag "modal"
      = turbo_frame_tag "snackbar" do
        = render("snackbar", message: flash[:notice]) if flash[:notice].present?
      = render "main_nav" if user_signed_in?
      = yield
```

The following is rendered:

<img width="510" alt="Screenshot 2023-10-07 at 21 17 48" src="https://github.com/haml/haml/assets/234051/877ac8f5-1d08-49da-8a8e-643d0074af98">

**After upgrading to Rails 7.1, it is rendered like this:**

<img width="880" alt="Screenshot 2023-10-07 at 21 23 09" src="https://github.com/haml/haml/assets/234051/bffddd77-be85-47c4-b3b5-133db6e7acb0">

Notice how the `turbo_frame_tag "snackbar"` contains a full render of the head and body again.

The same layout works fine written as `.erb`, so I think this has to do with Haml. It also works with the Haml layout as long as `flash[:notice]` evaluates to something truthy.

### Workaround

Changing the nesting like this makes everything work:

```haml
  = turbo_frame_tag "snackbar" do
    - if flash[:notice].present?
      = render "snackbar", message: flash[:notice]

```

**Comment by timdiggins** [comment]
I can confirm that this problem doesn't exist with haml-5 and rails 7.1 but not with haml-6.0+ and rails 7.1

Haml 5 (see https://github.com/timdiggins/haml-issue-1154/tree/issue-1153-haml-5 for repro with test)
![haml-5](https://github.com/haml/haml/assets/18395/b3b9c13e-c6a1-4e05-af35-545223a516be)

Haml 6 (https://github.com/timdiggins/haml-issue-1154/tree/issue-1153-haml-6.0.0 for repro with test)
![haml-6 0 0](https://github.com/haml/haml/assets/18395/016fac48-ad1f-4f3c-a8fd-eec4495af8cc)

**Comment by smdern** [comment]
I'm observing the same issue with haml v6 and rails v7.1. Haml v6 is not safe to be used with rails v7.1. The following commits https://github.com/rails/rails/pull/47194 and https://github.com/rails/rails/pull/45731 have changed the way captured output is buffered. 

Here is a succinct reproduction:

```haml
%h1 Some heading

= tag.div do
  = nil
```

<img width="215" alt="Screenshot 2023-10-12 at 13 08 27" src="https://github.com/haml/haml/assets/595639/4522b8e0-bd25-49f9-abda-751fbf641f05">

I'm unable to reproduce this issue when using erb.

It appears that many of the haml project's helper tests fail when running with rails 7.1. Here is a gist with the failing test output.

https://gist.github.com/aaronjensen/7a6607a25c0c6296981ccf83cc22a525

**Comment by antoineLeclercq** [comment]
I'm observing the same issue in my projects.

**Comment by aaronjensen** [comment]
This change appears to be the culprit, though I cannot say if it's doing something wrong or haml is: https://github.com/rails/rails/pull/47194/files?diff=unified&w=0#diff-dd3a604b68beb3b1a8c9b132a4b69a88606316c68b1c458266d308c3e2c04109L46-R49

**Comment by aaronjensen** [comment]
Fix in Rails coming: https://github.com/rails/rails/pull/49612

**Comment by byroot** [comment]
I merged the Rails fix and backported to `7-1-stable`. Please let me know if somehow that doesn't fix the bug. It should be released as part of rails `7.1.2`.

**Comment by manufaktor** [comment]
Thank you @byroot, works fine again with `7-1-stable `
